### PR TITLE
Draft NOTES for 2.0.0~RC1

### PR DESCRIPTION
**Pull request by AtnNn**
Ping @danielmewes @mglukhovsky

### CODE REVIEW

File: NOTES.md [hunk 1]
```diff
@@ -1,3 +1,186 @@
...
+Released on 2015-03-23
+
+This is a release candidate for the upcoming RethinkDB 2.0 release.
+It is not for production use. It might still have bugs. If you decide
+to test it, please [back up all][backup-docs] your data first.
+
+Release highlights:
+* Support for attaching a changefeed to the `get_all` and `union` commands
+* Improved support for asynchronous queries
+* Added a new feature called `tables` which can be used to store JSON documents
```
Comment (danielmewes) [inline NOTES.md]: It's a pretty useful feature, but also probably a place holder?

File: NOTES.md [hunk 2]
```diff
@@ -1,3 +1,186 @@
+# Release 2.0.0-RC1 (Musashi)
```
Comment (mglukhovsky) [inline NOTES.md]: Can we leave the release name off, as this is an RC release? (presumably it should be consistent with the actual 2.0 release name).

File: NOTES.md [hunk 3]
```diff
@@ -1,3 +1,186 @@
...
+* Improved support for asynchronous queries
+* Added a new feature called `tables` which can be used to store JSON documents
+
+Read the [blog post][2.0-RC1-blog] for more details.
+
+[2.0-RC1-blog]: http://rethinkdb.com/blog/2.0-release-candidate/
+
+## Compatibility ##
+
+This release may not be compatible with RethinkDB 2.0.
```
Comment (danielmewes) [inline NOTES.md]: We should make this more specific.
"Data files accessed with this release candidate might be incompatible with the final version of RethinkDB 2.0."

File: NOTES.md (hunk 1 above)
Comment (mglukhovsky) [inline NOTES.md]: Haha, this is a placeholder, right? :smile:

File: NOTES.md [hunk 4]
```diff
@@ -1,3 +1,186 @@
...
+`python -m rethinkdb._negative_zero_check` after upgrading both the
+server and Python driver.  See the output of `python -m
+rethinkdb._negative_zero_check --help` for additional options.
+
+[1.14-outdated-index]: http://rethinkdb.com/docs/troubleshooting/#my-secondary-index-is-outdated
+[backup-docs]: http://www.rethinkdb.com/docs/backup/
+
+### API-breaking changes ###
+
+* `between` no longer accepts `null` bounds. The new `r.minval` and
```
Comment (danielmewes) [inline NOTES.md]: I think there should be periods at the end of these bullet points.

File: NOTES.md [hunk 5]
```diff
@@ -1,3 +1,186 @@
+# Release 2.0.0-RC1 (Musashi)
+
+Released on 2015-03-23
+
+This is a release candidate for the upcoming RethinkDB 2.0 release.
+It is not for production use. It might still have bugs. If you decide
+to test it, please [back up all][backup-docs] your data first.
```
Comment (mglukhovsky) [inline NOTES.md]: Could we add a second sentence (new paragraph) asking people to report issues on GitHub or the Google Group, if they'd like to be involved with testing the RC?

File: NOTES.md (hunk 3 above)
Comment (mglukhovsky) [inline NOTES.md]: Could you elaborate on this? "may not be compatible" is a pretty broad statement.

File: NOTES.md [hunk 6]
```diff
@@ -1,3 +1,186 @@
...
+
+* `indexes_of` has been renamed to `offsets_of`
+
+* The `squash` argument to `changes` now defaults to `false`
+
+* The type hierarchy for exception types in the Python driver changed.
+  All exceptions including `RqlDriverError` now inherit from the
+  `RqlError` type.  `RqlRuntimeError`, `RqlCompileError` and
+  `RqlClientError` additionally inherit from the new `RqlQueryError`
+  type.
```
Comment (danielmewes) [inline NOTES.md]: There should be no period after the last senctence (replaces my previous comment about adding periods to the other points)

File: NOTES.md [hunk 7]
```diff
@@ -1,3 +1,186 @@
...
+
+## New Features ##
+
+* ReQL
+  * Added support for changefeeds on `get_all` and `union` queries (#3642)
+  * `between` no longer accepts `null` as a bound. The new `r.minval` and `r.maxval` can be used instead (#1023)
+  * Added support for getting the state of a changefeed using the new `include_states` optarg to `changes` (#3709)
+* Drivers
+  * Added support for non-blocking `cursor.get` (#3529)
```
Comment (danielmewes) [inline NOTES.md]: The function is called `next` not `get`

File: NOTES.md [hunk 8]
```diff
@@ -1,3 +1,186 @@
...
+  * Added support for parallel query execution on a single connection (#3754)
```
Comment (danielmewes) [inline NOTES.md]: To distinguish from the parallel execution of different parts of a single query, we should maybe rephrase this:
"Added support for executing multiple queries in parallel on a single connection"

File: NOTES.md [hunk 9]
```diff
@@ -1,3 +1,186 @@
...
+distinct values in primary and secondary indexes.  This could lead to
+inconsistent behavior and wrong query results in some rare cases.
+
+Starting with RethinkDB 2.0, -0 and +0 are indexed as equal values.
+Secondary indexes can be [migrated online][1.14-outdated-index] using
+the `rethinkdb index-rebuild` utility.  If any of your documents have
+negative zero values in their primary keys, those documents will
+become partially inaccessible in RethinkDB 2.0. You will need to
+re-import the affected tables using the `rethinkdb dump` and
+`rethinkdb restore` commands. See [Back up your data][backup-docs] for
```
Comment (mglukhovsky) [inline NOTES.md]: > See [Back up your data] for more information.

Odd capitalization. I would say:

See the article "[Back up your data]" for more information.

File: NOTES.md [hunk 10]
```diff
@@ -1,3 +1,186 @@
...
+* ReQL
+  * Added support for changefeeds on `get_all` and `union` queries (#3642)
+  * `between` no longer accepts `null` as a bound. The new `r.minval` and `r.maxval` can be used instead (#1023)
+  * Added support for getting the state of a changefeed using the new `include_states` optarg to `changes` (#3709)
+* Drivers
+  * Added support for non-blocking `cursor.get` (#3529)
+  * Added support for parallel query execution on a single connection (#3754)
+  * Consolidated the return types and use the new `ResponseNotes` field to convey extra information (#3715)
+* Python driver
+  * Add an optional script that warns for documents with negative zero as a primary key (#3637)
```
Comment (mglukhovsky) [inline NOTES.md]: Inconsistent: Add -> Added

Comment (danielmewes) [inline NOTES.md]: More precisely: "with negative zero in a primary key" (the primary key can be an array that has a negative zero as one of its elements)

File: NOTES.md [hunk 11]
```diff
@@ -1,3 +1,186 @@
...
+
+## Improvements ##
+
+* Server
+  * Report open cursors as a single entry in the jobs table (#3662)
+  * No longer send timestamps between servers in `batchspec_t` (#2671)
```
Comment (mglukhovsky) [inline NOTES.md]: Timestamps are no longer sent between servers in `batchspec_t`.

File: NOTES.md [hunk 12]
```diff
@@ -1,3 +1,186 @@
...
+  * No longer perform some expensive checks in release mode (#3656)
```
Comment (mglukhovsky) [inline NOTES.md]: Some expensive checks are no longer performed in release mode.

File: NOTES.md [hunk 13]
```diff
@@ -1,3 +1,186 @@
...
+  * Include the remote port number in the heartbeat timeout message (#2891)
+  * Improved the ordering and throttling of reads and writes (#1606)
+  * Limit the default number of documents per batch to reduce latency (#3806)
+  * Execute multiple queries in parallel on a single connection (#3296)
+  * Improved the performance of sending responses (#3744)
+* Web UI
+  * Added a modifiable query limit (#3910)
+  * Add an "add table" button to each database (#3522)
```
Comment (mglukhovsky) [inline NOTES.md]: Inconsistent: Add -> Added

File: NOTES.md [hunk 14]
```diff
@@ -1,3 +1,186 @@
...
+
+## Improvements ##
+
+* Server
+  * Report open cursors as a single entry in the jobs table (#3662)
+  * No longer send timestamps between servers in `batchspec_t` (#2671)
+  * No longer perform some expensive checks in release mode (#3656)
+  * Include the remote port number in the heartbeat timeout message (#2891)
+  * Improved the ordering and throttling of reads and writes (#1606)
+  * Limit the default number of documents per batch to reduce latency (#3806)
```
Comment (danielmewes) [inline NOTES.md]: This applies specifically to write queries:
"Limit the number of documents per write batch to reduce the impact of large writes on other queries"

File: NOTES.md [hunk 15]
```diff
@@ -1,3 +1,186 @@
...
+  * Improved the performance of sending responses (#3744)
+* Web UI
+  * Added a modifiable query limit (#3910)
+  * Add an "add table" button to each database (#3522)
+* ReQL
+  * `table.rebalance` with insufficient data is no longer an error (#3679)
+  * Renamed `indexes_of` to `offsets_of` to avoid confusion with secondary indexes (#3265)
+  * Removed `any` and `all` in favor of `or` and `and` (#1581)
+  * Immediately return an empty first batch when the result is a changefeed (#3852)
+  * No longer include trivial changes from `return_changes` (#3697)
```
Comment (mglukhovsky) [inline NOTES.md]: Trivial changes are filtered out from `return_changes`.

File: NOTES.md [hunk 16]
```diff
@@ -1,3 +1,186 @@
...
+  * Reduce the size of profiles (#3218)
```
Comment (mglukhovsky) [inline NOTES.md]: Reduce -> Reduced

File: NOTES.md [hunk 17]
```diff
@@ -1,3 +1,186 @@
...
+  * Report open cursors as a single entry in the jobs table (#3662)
+  * No longer send timestamps between servers in `batchspec_t` (#2671)
+  * No longer perform some expensive checks in release mode (#3656)
+  * Include the remote port number in the heartbeat timeout message (#2891)
+  * Improved the ordering and throttling of reads and writes (#1606)
+  * Limit the default number of documents per batch to reduce latency (#3806)
+  * Execute multiple queries in parallel on a single connection (#3296)
+  * Improved the performance of sending responses (#3744)
+* Web UI
+  * Added a modifiable query limit (#3910)
```
Comment (danielmewes) [inline NOTES.md]: "Added a configurable limit for the results per page in the Data Explorer"

File: NOTES.md [hunk 18]
```diff
@@ -1,3 +1,186 @@
...
+  * Add an "add table" button to each database (#3522)
+* ReQL
+  * `table.rebalance` with insufficient data is no longer an error (#3679)
+  * Renamed `indexes_of` to `offsets_of` to avoid confusion with secondary indexes (#3265)
+  * Removed `any` and `all` in favor of `or` and `and` (#1581)
+  * Immediately return an empty first batch when the result is a changefeed (#3852)
+  * No longer include trivial changes from `return_changes` (#3697)
+  * Reduce the size of profiles (#3218)
+  * No longer squash changefeeds by default (#3904)
```
Comment (mglukhovsky) [inline NOTES.md]: Changefeeds are no longer squashed by default.

File: NOTES.md [hunk 19]
```diff
@@ -1,3 +1,186 @@
...
+* JavaScript driver
+  * Added an upper bound to the bluebird dependency (#3823)
+* Ruby driver
+  * Added a `timeout` option to `r.connect` (#1666)
+  * Improved the code style (#3900, #3901, #3906)
+  * Allow strings as keys in the config options (#3905)
```
Comment (mglukhovsky) [inline NOTES.md]: Strings are now allowed as keys in the config options.

File: NOTES.md [hunk 20]
```diff
@@ -1,3 +1,186 @@
...
+* Build
+  * Upgraded to a more recent version of V8 and drop support for out-of-tree V8 (#3472)
```
Comment (mglukhovsky) [inline NOTES.md]: Upgraded to a more recent version of V8 and dropped support for out-of-tree V8

File: NOTES.md [hunk 21]
```diff
@@ -1,3 +1,186 @@
...
+  * Execute multiple queries in parallel on a single connection (#3296)
+  * Improved the performance of sending responses (#3744)
+* Web UI
+  * Added a modifiable query limit (#3910)
+  * Add an "add table" button to each database (#3522)
+* ReQL
+  * `table.rebalance` with insufficient data is no longer an error (#3679)
+  * Renamed `indexes_of` to `offsets_of` to avoid confusion with secondary indexes (#3265)
+  * Removed `any` and `all` in favor of `or` and `and` (#1581)
+  * Immediately return an empty first batch when the result is a changefeed (#3852)
```
Comment (danielmewes) [inline NOTES.md]: I would put this under "server" and write "send back" instead of "return".

The empty batch is completely invisible to the user.

File: NOTES.md [hunk 22]
```diff
@@ -1,3 +1,186 @@
...
+  * Improved the code style (#3900, #3901, #3906)
+  * Allow strings as keys in the config options (#3905)
+* Build
+  * Upgraded to a more recent version of V8 and drop support for out-of-tree V8 (#3472)
+  * Added support for building with Python 3 (#3731)
+* Packaging
+  * Got rid of the outdated bash completion script (#719)
+  * Allow installing RethinkDB in 32-bit OS X on a 64-bit processor (#1595)
+* Tests
+  * Increase the number of retries in the `RDBBtree` tests to avoid false positives (#3805)
```
Comment (mglukhovsky) [inline NOTES.md]: Increase -> Increased

File: NOTES.md [hunk 23]
```diff
@@ -1,3 +1,186 @@
...
+
+## Bug Fixes ##
+
+* Server
+  * Fixed a race condition that could be caused by concurrent queries (#3766)
+  * No longer count deleted servers and tables during version check (#3692)
```
Comment (mglukhovsky) [inline NOTES.md]: Deleted servers and tables are no longer counted during version checks.

File: NOTES.md [hunk 24]
```diff
@@ -1,3 +1,186 @@
...
+  * Made JSON parsing more strict (#3810)
+  * Fixed a bug that could cause the server to crash when killed (#3792)
+  * No longer allow renaming a database to "rethinkdb" (#3858)
```
Comment (mglukhovsky) [inline NOTES.md]: Databases can no longer be renamed to "rethinkdb".

File: NOTES.md [hunk 25]
```diff
@@ -1,3 +1,186 @@
...
+  * Return an initial value for point changefeeds on system tables (#3723)
+  * Improve the handling of negative zero (#3637)
```
Comment (mglukhovsky) [inline NOTES.md]: Improve -> Improved

File: NOTES.md [hunk 26]
```diff
@@ -1,3 +1,186 @@
...
+  * Correctly abort `order_by.limit` changefeeds when a table because unavailable (#3932)
```
Comment (danielmewes) [inline NOTES.md]: because -> becomes

**Comment by mglukhovsky** [comment]
@AtnNn @danielmewes, added my notes.

**Comment by danielmewes** [comment]
Added mine too.

**Comment by AtnNn** [comment]
Thanks for the feedback @mglukhovsky @danielmewes. I have integrated all your suggestions.

File: NOTES.md [hunk 27]
```diff
@@ -1,3 +1,190 @@
+# Release 2.0.0-RC1
+
+Released on 2015-03-23
+
+This is a release candidate for the upcoming RethinkDB 2.0 release.
+It is not for production use. It might still have bugs. If you decide
+to test it, please [back up all][backup-docs] your data first.
+
+Please report any bugs you find on github
```
Comment (mglukhovsky) [inline NOTES.md]: Capitalize GitHub.

File: NOTES.md (hunk 5 above)
Comment (mglukhovsky) [inline NOTES.md]: Thanks, done!

**Comment by mglukhovsky** [comment]
@AtnNn: great! One minor capitalization fix.

**Comment by AtnNn** [comment]
Thanks @mglukhovsky
